### PR TITLE
Fix #491, updating rendering_config()

### LINKED ISSUE #491: Rendering config extensibility
Because the rendering config is empty, its size is not known and decoders cannot skip over it. Because it is not backwards compatible, we could remove it entirely from this version of the spec. A version of the specification could use a new OBU.

### PR DESCRIPTION
**Pull request by sunghee-hwang**
To resolve the discord of two HRTFs

***
<a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/pull/509.html" title="Last updated on Jun 27, 2023, 10:55 PM UTC (7e0aa0a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/AOMediaCodec/iamf/509/3318721...7e0aa0a.html" title="Last updated on Jun 27, 2023, 10:55 PM UTC (7e0aa0a)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
 ### Rendering Config Syntax and Semantics ### {#syntax-rendering-config}
 
-During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]]. In this version of the specification, no additional metadata is required to configure the renderers, and as such, <code>rendering_config()</code> has an empty payload.
+During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]].
 
 <b>Syntax</b>
 
 ```
 class rendering_config() {
+  unsigned int (1) rendering_to_loudspeaker;
```
Comment (sunghee-hwang) [inline index.bs:1111]:
```suggestion
  unsigned int (1) rendering_to_stereo;
```

File: index.bs [hunk 2]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
 ```
 class rendering_config() {
+  unsigned int (1) rendering_to_stereo;
+  unsigned int (7) reserved;
 }
 ```
 
 <b>Semantics</b>
 
+<dfn noexport>rendering_to_stereo</dfn> indicates whether or not the channel-based [=Audio Element=] is rendered to stereo loudspeakers.
```
Comment (sunghee-hwang) [inline index.bs:1118]:
```suggestion
<dfn noexport>rendering_to_stereo</dfn> indicates whether or not the input [=Audio Element=] is rendered to stereo loudspeakers.
```

File: index.bs [hunk 3]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
 ```
 class rendering_config() {
+  unsigned int (1) rendering_to_stereo;
+  unsigned int (7) reserved;
 }
 ```
 
 <b>Semantics</b>
 
+<dfn noexport>rendering_to_stereo</dfn> indicates whether or not the input [=Audio Element=] is rendered to stereo loudspeakers.
```
Comment (felicialim) [inline index.bs:1118]:
```suggestion
<dfn noexport>rendering_to_stereo</dfn> indicates whether the input channel-based [=Audio Element=] is rendered to stereo loudspeakers or spatialized with a binaural renderer when played back on headphones. If the playback layout is a loudspeaker layout, or the [=audio_element_type=] is SCENE_BASED, the parsers SHALL ignore this field.
```

File: index.bs [hunk 4]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
+- 0: It indicates that the input [=Audio Element=] is rendered to the playback layout.
```
Comment (felicialim) [inline index.bs:1119]:
```suggestion
- 0: It indicates that the input [=Audio Element=] is rendered using the pre-defined renderer used for headphones.
```

File: index.bs [hunk 5]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
+- 1: It indicates that the input [=Audio Element=] is rendered to [=loudspeaker_layout=] = Stereo. If the playback layout is loudspeaker or for scene-based [=Audio Element=], parsers SHALL ignore it.
```
Comment (felicialim) [inline index.bs:1120]:
```suggestion
- 1: It indicates that the input [=Audio Element=] is rendered using the pre-defined renderer used for [=loudspeaker_layout=] = Stereo.
```

File: index.bs (hunk 4 above)
Comment (sunghee-hwang) [reply index.bs:1119]: > 
How about?
```suggestion
- 0: It indicates that the input [=Audio Element=] is rendered to binaural output.
```

Comment (felicialim) [reply index.bs:1119]: lgtm

File: index.bs (hunk 5 above)
Comment (felicialim) [reply index.bs:1120]: lgtm

**Review by felicialim — APPROVED** [review]

File: index.bs [hunk 6]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
 ### Rendering Config Syntax and Semantics ### {#syntax-rendering-config}
 
-During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]]. In this version of the specification, no additional metadata is required to configure the renderers, and as such, <code>rendering_config()</code> has an empty payload.
+During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]].
 
 <b>Syntax</b>
 
 ```
 class rendering_config() {
+  unsigned int (1) rendering_to_stereo;
```
Comment (cconcolato) [inline index.bs:1111]: Do we expect that future versions will still be using this field? If not, shouldn't this be an enum?

Comment (felicialim) [reply index.bs:1111]:
```suggestion
class rendering_config() {
  if (audio_element_type == CHANNEL_BASED) {
    unsigned int (2) headphones_rendering_mode;
    unsigned int (6) reserved;
  } 
}
```

where headphones_rendering_mode indicates how an input channel-based [=Audio Element=] SHALL be rendered when played back on headphones. If the playback layout is a loudspeaker layout, parsers SHALL ignore this field.

The enum values for `headphones_rendering_mode`: 
0: STEREO: the input audio element is rendered to loudspeaker_layout = Stereo
1: BINAURAL: the input audio element is rendered to binaural output.

(We should address the extensibility too. I'll create a separate issue for it)

File: index.bs [hunk 7]
```diff
@@ -1102,17 +1102,29 @@ class mix_presentation_element_annotations() {
...
+    unsigned int (6) reserved;
+  } 
+}
+  unsigned int (7) reserved;
 }
 ```
 
 <b>Semantics</b>
 
+<dfn noexport>rendering_to_stereo</dfn> indicates whether the input channel-based [=Audio Element=] is rendered to stereo loudspeakers or spatialized with a binaural renderer when played back on headphones. If the playback layout is a loudspeaker layout, or the [=audio_element_type=] is SCENE_BASED, the parsers SHALL ignore this field.
```
Comment (sunghee-hwang) [inline index.bs:1123]:
```suggestion
<dfn noexport>headphones_rendering_mode</dfn> indicates whether the input channel-based [=Audio Element=] is rendered to stereo loudspeakers or spatialized with a binaural renderer when played back on headphones. If the playback layout is a loudspeaker layout, or the [=audio_element_type=] is SCENE_BASED, the parsers SHALL ignore this field.
```

File: index.bs [hunk 8]
```diff
@@ -1102,17 +1102,29 @@ class mix_presentation_element_annotations() {
...
 
 ```
 class rendering_config() {
+class rendering_config() {
+  if (audio_element_type == CHANNEL_BASED) {
+    unsigned int (2) headphones_rendering_mode;
+    unsigned int (6) reserved;
+  } 
+}
+  unsigned int (7) reserved;
```
Comment (sunghee-hwang) [inline index.bs:1117]:
```suggestion
```

File: index.bs [hunk 9]
```diff
@@ -1102,17 +1102,29 @@ class mix_presentation_element_annotations() {
...
 }
 ```
 
 <b>Semantics</b>
 
+<dfn noexport>rendering_to_stereo</dfn> indicates whether the input channel-based [=Audio Element=] is rendered to stereo loudspeakers or spatialized with a binaural renderer when played back on headphones. If the playback layout is a loudspeaker layout, or the [=audio_element_type=] is SCENE_BASED, the parsers SHALL ignore this field.
+- 0: It indicates that the input [=Audio Element=] is rendered to binaural output.
+- 1: It indicates that the input [=Audio Element=] is rendered to [=loudspeaker_layout=] = Stereo. 
+
```
Comment (sunghee-hwang) [inline index.bs:1126]:
```suggestion
- 0: It indicates that the input [=Audio Element=] is rendered to [=loudspeaker_layout=] = Stereo. 
- 1: It indicates that the input [=Audio Element=] is rendered to binaural output.
```

**Review by sunghee-hwang — COMMENTED** [review]
More clarification

File: index.bs [hunk 10]
```diff
@@ -1102,17 +1102,27 @@ class mix_presentation_element_annotations() {
...
-During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]]. In this version of the specification, no additional metadata is required to configure the renderers, and as such, <code>rendering_config()</code> has an empty payload.
+During playback, an [=Audio Element=] SHOULD be rendered using a pre-defined renderer according to [[#processing-mixpresentation-rendering]].
 
 <b>Syntax</b>
 
 ```
 class rendering_config() {
+class rendering_config() {
+  if (audio_element_type == CHANNEL_BASED) {
+    unsigned int (2) headphones_rendering_mode;
```
Comment (tdaede) [inline index.bs:1112]: I see you switched this to (2), but you only specify values for 0 and 1. I would suggest either switching this back to (1), or specifying what to do for the currently reserved values of 2/3.

File: index.bs [hunk 11]
```diff
@@ -1102,17 +1102,24 @@ class mix_presentation_element_annotations() {
...
 
 <b>Syntax</b>
 
 ```
 class rendering_config() {
-}
+  if (audio_element_type == CHANNEL_BASED) {
+    unsigned int (2) headphones_rendering_mode;
+    unsigned int (6) reserved;
+  }
```
Comment (tdaede) [inline index.bs:1114]: We need to handle the else {}

File: index.bs [hunk 12]
```diff
@@ -1102,17 +1102,34 @@ class mix_presentation_element_annotations() {
...
 
 <b>Semantics</b>
 
+<dfn noexport>headphones_rendering_mode</dfn> indicates whether the input channel-based [=Audio Element=] is rendered to stereo loudspeakers or spatialized with a binaural renderer when played back on headphones. If the playback layout is a loudspeaker layout or the input [=Audio Element=] is not CHANNEL_BASED, the parsers SHALL ignore this field.
+
+- 0: It indicates that the input [=Audio Element=] is rendered to [=loudspeaker_layout=] = Stereo. 
+- 1: It indicates that the input [=Audio Element=] is rendered to binaural output.
+- 2~3: Reserved.
+
+Parsers encountering a Reserved value SHALL ignore the Mix Presentation OBU that contains this [=rendering_config()=].
```
Comment (tdaede) [inline index.bs:1126]: "Reserved value of headphones_rendering_mode"

**Pull request merged**